### PR TITLE
COH-50 : Cohort resources cannot be voided

### PR DESCRIPTION
**Pull request by ManojLL**
### Issue worked on:
https://issues.openmrs.org/browse/COH-50?filter=-1

### Description

This PR fixed the issue in  [(COH-50)](https://issues.openmrs.org/browse/COH-50?filter=-1) DELETE Cohort endpoint.

### CODE REVIEW

File: api/src/main/java/org/openmrs/module/cohort/api/impl/CohortTypeServiceImpl.java
Comment (Piumal1999) [inline api/src/main/java/org/openmrs/module/cohort/api/impl/CohortTypeServiceImpl.java:61]: I tested this locally and found that here (in cohortTypes), it voids the cohortType automatically even if we haven't manually set the fields related to voiding. 

But in cohorts, it doesn't automatically void the cohort unless we manually set the fields related to voiding.

**Review by Piumal1999 — COMMENTED** [review]
Hi @ibacher, could you please review this PR? I added some comments as well.

**Comment by dkayiwa** [comment]
@ManojLL can you add some tests for this?

**Comment by ManojLL** [comment]
@dkayiwa I added some tests for voidCohort method. Can you review?

File: api/src/main/java/org/openmrs/module/cohort/api/impl/CohortServiceImpl.java
Comment (ibacher) [inline api/src/main/java/org/openmrs/module/cohort/api/impl/CohortServiceImpl.java]: We don't use `*` imports. Please configure your IDE accordingly.

File: api/src/main/java/org/openmrs/module/cohort/api/impl/CohortTypeServiceImpl.java
Comment (ibacher) [reply api/src/main/java/org/openmrs/module/cohort/api/impl/CohortTypeServiceImpl.java:61]: Theoretically, there's a VoidHandler AOP that's supposed to do this for most cases... Why it isn't working for Cohorts is a good question, though...

**Review by ibacher — CHANGES_REQUESTED** [review]

**Comment by ManojLL** [comment]
@ibacher @dkayiwa  can you review this PR?

File: api/src/main/java/org/openmrs/module/cohort/api/impl/CohortServiceImpl.java
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/module/cohort/api/impl/CohortServiceImpl.java]: Is this needed?

File: api/src/test/java/org/openmrs/module/cohort/api/impl/CohortServiceImplTest.java
Comment (dkayiwa) [inline api/src/test/java/org/openmrs/module/cohort/api/impl/CohortServiceImplTest.java]: Can we use the original formatting?

File: api/src/main/java/org/openmrs/module/cohort/api/impl/CohortServiceImpl.java
Comment (ManojLL) [reply api/src/main/java/org/openmrs/module/cohort/api/impl/CohortServiceImpl.java]: cohort won't voided automatically. That is why manually set voided true.

Comment (dkayiwa) [reply api/src/main/java/org/openmrs/module/cohort/api/impl/CohortServiceImpl.java]: Do you have a test to demonstrate it?

Comment (ManojLL) [reply api/src/main/java/org/openmrs/module/cohort/api/impl/CohortServiceImpl.java]: Here I attached test screenshots.
Screenshot from 2023-03-20 15-03-37
Screenshot from 2023-03-20 15-03-54

Comment (dkayiwa) [reply api/src/main/java/org/openmrs/module/cohort/api/impl/CohortServiceImpl.java]: I mean tests in this pull request. Not a screenshot.

Comment (dkayiwa) [reply api/src/main/java/org/openmrs/module/cohort/api/impl/CohortServiceImpl.java]: In other words, tests which would fail if you removed that `for loop`

Comment (ManojLL) [reply api/src/main/java/org/openmrs/module/cohort/api/impl/CohortServiceImpl.java]: > In other words, tests which would fail if you removed that `for loop`

If I remove `for loop` tests wont fail

Comment (dkayiwa) [reply api/src/main/java/org/openmrs/module/cohort/api/impl/CohortServiceImpl.java]: Which means that you do not have tests to demonstrate the problem that you claim to require the `for loop`

Comment (ManojLL) [reply api/src/main/java/org/openmrs/module/cohort/api/impl/CohortServiceImpl.java]: @dkayiwa  I updated the test can you review ?

File: api/src/main/java/org/openmrs/module/cohort/api/impl/CohortServiceImpl.java
Comment (ibacher) [inline api/src/main/java/org/openmrs/module/cohort/api/impl/CohortServiceImpl.java]: Is there a reason we need these two explicit `setVoided()` and `setVoidReason()` calls rather than calling the service method?

File: api/src/test/java/org/openmrs/module/cohort/api/impl/CohortServiceImplTest.java
Comment (ibacher) [inline api/src/test/java/org/openmrs/module/cohort/api/impl/CohortServiceImplTest.java]: It would be better to write a separate test for this. The test here isn't really valid because ultimately we mock the `createOrUpdate()` call...

File: api/src/main/java/org/openmrs/module/cohort/api/impl/CohortServiceImpl.java
Comment (ManojLL) [reply api/src/main/java/org/openmrs/module/cohort/api/impl/CohortServiceImpl.java]: @ibacher I thought that, it would be better to call the two explicit would be better than calling `Context.getService` multiple times inside the `for loop` considering the performance, that is why I use `setVoided` and `setVoidReasin` methods.

File: api/src/test/java/org/openmrs/module/cohort/api/impl/CohortServiceImplTest.java
Comment (ManojLL) [reply api/src/test/java/org/openmrs/module/cohort/api/impl/CohortServiceImplTest.java]: @ibacher This test is to check whether the child `cohortMember` is voided when the parent cohort is deleted. So I added the `cohortMember` to the `cohort` and used assertions to check the `cohort` and `cohortMember`  is voided after deleting the `cohort`

File: api/src/main/java/org/openmrs/module/cohort/api/impl/CohortTypeServiceImpl.java
Comment (dkayiwa) [reply api/src/main/java/org/openmrs/module/cohort/api/impl/CohortTypeServiceImpl.java:61]: For the voiding to happen automatically, the method should be renamed to `voidCohortM` because of this: https://github.com/openmrs/openmrs-core/blob/2.6.0/api/src/main/java/org/openmrs/aop/RequiredDataAdvice.java#L191-L218

**Comment by dkayiwa** [comment]
@ManojLL the method needs renaming to `voidCohortM` for this to work automatically as seen here: https://github.com/openmrs/openmrs-core/blob/2.6.0/api/src/main/java/org/openmrs/aop/RequiredDataAdvice.java#L191-L218

If we want other platform AOP things to work for other methods like saveCohort, etc, we would need to rename those methods too. But any way, that would be a different ticket.

File: api/src/main/java/org/openmrs/module/cohort/api/impl/CohortServiceImpl.java
Comment (ibacher) [reply api/src/main/java/org/openmrs/module/cohort/api/impl/CohortServiceImpl.java]: I'd tend to prefer the service call. You could always do:

```
CohortMemberService cms = Context.getService(CohortMemberService.class);
// loop
   cms.voidCohortMember();
// end loop
```

That way, it's the CohortMemberService's responsibility to know how to void a cohort member (and also it fixes the real problem which is that calling these methods depends on Hibernate's implicit context to flush those changes to the DB whereas calling the service method will actually ensure that those changes are always written.

File: api/src/test/java/org/openmrs/module/cohort/api/impl/CohortServiceImplTest.java
Comment (ibacher) [reply api/src/test/java/org/openmrs/module/cohort/api/impl/CohortServiceImplTest.java]: So, what I was trying to get at is the way in which the mock changes this from how the code would run in production because the key thing about voiding the memberships is that the voiding gets written to the database, but because we've mocked the call to `createOrUpdate()` nothing is getting written to a database, so, ultimately, this test ends up verifying that `cohortMember.setVoided()` was called, but not that the `cohortMember` was actually saved in a voided state.

File: api/src/main/java/org/openmrs/module/cohort/api/impl/CohortServiceImpl.java
Comment (ManojLL) [reply api/src/main/java/org/openmrs/module/cohort/api/impl/CohortServiceImpl.java]: @ibacher After rename the `voidChort` method to `voidCohortM` ,no need to call `cohortMemberservice` void method to make `cohortmember` to voided. when delete a `cohort` , the child `cohortmember` getting Voided.

**Comment by ManojLL** [comment]
@dkayiwa rename the` voidCohort` metod to `voidcohortM`, Can you review the changes?

**Comment by dkayiwa** [comment]
@ManojLL do you think you can add a test?

**Comment by ManojLL** [comment]
> @ManojLL do you think you can add a test?

@dkayiwa I tried to write test for this method, but unable to find a way to check whether the `cohort` or `cohortMember` is voided after mocking `voidCohortM` method.

**Comment by dkayiwa** [comment]
@ManojLL would it be easier for you to just put un mocked tests here? https://github.com/openmrs/openmrs-module-cohort/blob/master/api/src/test/java/org/openmrs/module/cohort/api/dao/CohortGenericDaoTest.java

**Comment by ManojLL** [comment]
> @ManojLL would it be easier for you to just put un mocked tests here? https://github.com/openmrs/openmrs-module-cohort/blob/master/api/src/test/java/org/openmrs/module/cohort/api/dao/CohortGenericDaoTest.java

@dkayiwa I added tests for `voidcohortM` method and `voidCohortType` method, you can review.

**Comment by dkayiwa** [comment]
I do not see your test calling the `service.voidCohortM` method of interest.

**Comment by ManojLL** [comment]
@dkayiwa I updated the test

**Comment by dkayiwa** [comment]
The test that you added passes even when you do not fix the reported problem.
Can you use the real service without mocking it?

**Comment by ManojLL** [comment]
@dkayiwa updated the tests, You can review.

File: api/src/test/java/org/openmrs/module/cohort/api/dao/CohortGenericDaoTest.java
Comment (dkayiwa) [inline api/src/test/java/org/openmrs/module/cohort/api/dao/CohortGenericDaoTest.java]: Why are you setting the void properties of the cohort and then at the same time call the voidCohort service method? It is the service method that is supposed to take care of that.

Comment (ManojLL) [reply api/src/test/java/org/openmrs/module/cohort/api/dao/CohortGenericDaoTest.java]: @dkayiwa I thought that, the `cohort` which is passing to the `voidCoghortM` method needs be voided true.
btw I updated test.

File: api/src/test/java/org/openmrs/module/cohort/api/dao/CohortGenericDaoTest.java
Comment (dkayiwa) [inline api/src/test/java/org/openmrs/module/cohort/api/dao/CohortGenericDaoTest.java]: Why are you calling `dao.createOrUpdate()` after `service.voidCohortM()`? Isn't the service method the one supposed to take care of the voiding?

Comment (ManojLL) [reply api/src/test/java/org/openmrs/module/cohort/api/dao/CohortGenericDaoTest.java]: @dkayiwa I just call `dao.createOrUpdate` because it has used in `voidCohortM` method.

File: api/src/test/java/org/openmrs/module/cohort/api/dao/CohortGenericDaoTest.java
Comment (dkayiwa) [inline api/src/test/java/org/openmrs/module/cohort/api/dao/CohortGenericDaoTest.java:155]: Can you assert that the cohort is not voided before calling service.voidCohort?

File: api/src/test/java/org/openmrs/module/cohort/api/dao/CohortGenericDaoTest.java
Comment (dkayiwa) [inline api/src/test/java/org/openmrs/module/cohort/api/dao/CohortGenericDaoTest.java:157]: After calling service.voidCohort, can you fetch the cohort again from the dao and then assert that it is voided?

File: api/src/test/java/org/openmrs/module/cohort/api/dao/CohortGenericDaoTest.java
Comment (ManojLL) [reply api/src/test/java/org/openmrs/module/cohort/api/dao/CohortGenericDaoTest.java:155]: @dkayiwa i updated test to assert that the cohort is not voided before calling to service.voidCohortM.

**Comment by Piumal1999** [comment]
Thank you all for getting this done.

@dkayiwa @ibacher will it be possible to do a release with this fix? 
Because the error with the `delete cohort` is a blocker for completing this PR: https://github.com/openmrs/openmrs-esm-patient-management/pull/578.

**Comment by dkayiwa** [comment]
@Piumal1999 feel empowered to do the release by yourself. Here are the instructions: https://wiki.openmrs.org/display/docs/Releasing+a+Module+from+Bamboo

**Comment by Piumal1999** [comment]
Sure, I would love to. I just created a help desk ticket to get access to the bamboo dashboard. 
Thank you @dkayiwa

**Comment by ibacher** [comment]
I’d prefer if we hold off on releasing this just yet. There’s no reason we need to do that to unblock the issue; we can just update the O3 release to use the snapshot and then, once we inevitably find another problem or two, we can fix things without all the hassle. We’ll cut a release of this the next time we need to release the 3.x RefApp (which will likely be soon anyways).

**Comment by Piumal1999** [comment]
> I’d prefer if we hold off on releasing this just yet. There’s no reason we need to do that to unblock the issue; we can just update the O3 release to use the snapshot and then, once we inevitably find another problem or two, we can fix things without all the hassle. We’ll cut a release of this the next time we need to release the 3.x RefApp (which will likely be soon anyways).

Oh sorry @ibacher. I already did the release. 

> we can just update the O3 release to use the snapshot

Btw, how do we change the module versions used in dev3? Is it done by updating [this](https://github.com/openmrs/openmrs-distro-referenceapplication/blob/main/distro/pom.xml)?

**Comment by dkayiwa** [comment]
> Is it done by updating [this](https://github.com/openmrs/openmrs-distro-referenceapplication/blob/main/distro/pom.xml)?

Yes.

**Pull request merged**
